OpenOversight/tests/conftest.py

```diff
@@ -300,30 +300,27 @@ def db(app):
     with app.app_context():
         _db.app = app
         _db.create_all()
-        connection = _db.engine.connect()
-        session = scoped_session(session_factory=sessionmaker(bind=connection))
-        _db.session = session
-        add_mockdata(session)
-        session.commit()
-        connection.close()
-        session.remove()
+        with _db.engine.begin() as connection:
+            _db.session = scoped_session(sessionmaker(bind=connection))
+            add_mockdata(_db.session)
+            _db.session.remove()
 
         yield _db
 
 
 @pytest.fixture(scope="function")
 def session(db):
     """Creates a new database session for a test."""
-    with db.engine.connect() as connection:
-        with connection.begin():
-            session = scoped_session(session_factory=sessionmaker(bind=connection))
-            db.session = session
-
-            try:
-                yield session
-            finally:
-                session.remove()
-                connection.close()
+    with db.engine.connect() as connection, connection.begin() as tx:
+        db.session = scoped_session(sessionmaker(bind=connection))
+
+        try:
+            yield db.session
+        finally:
+            db.session.remove()
+            # If an error was raised during test, tx is already rolled back
+            if tx.is_active:
+                tx.rollback()
```

Comment:
We were getting the warning `SAWarning: transaction already deassociated from connection` from 2 unit tests which intentionally violated database constraints, causing the transaction to be automatically rolled back when the commit failed. Then, when the fixture was being cleaned up, it was trying to roll back the transaction a second time.

Instead, this change checks whether the transaction is still active and rolls back if needed.